Patient Builder Improvements

DPC15038 opened the pull request.

thecristen commented on app/assets/javascripts/templates/patient_builder/edit_criteria.hbs:
```diff
@@ -21,6 +21,45 @@
+      {{/ifCond}}          
+      {{!=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=}}
+      {{#ifCond result_information.length ">" 0}}
+        <div style="line-height:135%">&nbsp;</div> {{!Adds a thin blank line (thinner than <br>) between each item}}
```
An `<hr>` might make more sense than a pseudo-empty `<div>` here. I'd also avoid using the `style` attribute here and add a CSS class to adjust the appearance.

thecristen commented on app/assets/javascripts/templates/patient_builder/edit_criteria.hbs:
```diff
@@ -21,6 +21,45 @@
+      <div class="row">
+        <div class="col-md-12">
+      {{!The following ifConds display information when the window is minimized}}
+      {{#ifCond fulfillment_information.length ">" 0}}
```
Would `{{#if fulfillment_information}}` work just as well here?

thecristen commented on app/assets/javascripts/templates/patient_builder/edit_criteria.hbs:
```diff
@@ -21,6 +21,45 @@
+      {{!=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=}}
+      {{#ifCond field_value_information.length ">" 0}}
+        <div style="line-height:135%">&nbsp;</div> {{!Adds a thin blank line (thinner than <br>) between each item}}
+        <label class="control-label"><i class="fa fa-bar-chart-o" aria-hidden="true"></i>&nbsp;Field Values<input class="sr-only" disabled></label>
```
Secret font-awesome tip: adding the `fa-fw` class to the icon makes it fixed width, which will reduce the need to include `&nbsp;` after every icon.

thecristen commented on app/assets/javascripts/templates/patient_builder/edit_criteria.hbs:
```diff
@@ -21,6 +21,45 @@
+        <div style="line-height:135%">&nbsp;</div> {{!Adds a thin blank line (thinner than <br>) between each item}}
+        <label class="control-label"><i class="fa fa-bar-chart-o" aria-hidden="true"></i>&nbsp;Field Values<input class="sr-only" disabled></label>
+          {{#each field_value_information}}
+          <div> <strong> &ndash; </strong> &nbsp; {{this}} </div> 
```
This is more of a preference thing, but maybe replace`<strong> &ndash; </strong> &nbsp;` with `<i class="fa fa-fw fa-minus" aria-hidden="true"></i>` (or any other icon)? [Using list icons might work well here too](http://fontawesome.io/examples/#list).

thecristen commented on app/assets/javascripts/views/patient_builder/data_criteria.js.coffee:
```diff
@@ -95,6 +95,68 @@ class Thorax.Views.EditCriteriaView extends Thorax.Views.BuilderChildView
+          #format the QRDA(or whatever healthcare language this is written in) by replacing underscores with spaces and capitalizing each word
+          
+          fulfillments_information_array.push(that)
+          #Add this formatted string to an array containing all the other fulfillment values
```
I notice that a lot of this is building an array of strings corresponding to an array of things, and I see a great opportunity to use the map function here. Map basically takes an array, performs a function on every item in the array, and returns a new array. For instance, I think replacing the above section with this should produce the same result:

```coffeescript
if @model.attributes.references?
  references_information_array = @model.attributes.references.map((r) ->
     # using .get to retreive attributes since r is a Thorax object
     if r.get('reference_type')
       return r.get('reference_type').replace(/_/g, ' ').replace(/\w\S*/g, (txt) ->  txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase()) + ": " + r.get('description') + ": " + r.get('start_date')
     else
       return r.get('description') + ": " + r.get('start_date'))
```

DPC15038 commented:
![screen shot 2016-06-08 at 1 48 43 pm](https://cloud.githubusercontent.com/assets/17532163/15904659/dc1a2982-2d7f-11e6-990e-5e9e5e9c85be.png)

rutiah commented:
Really good! Is there a way to reduce the amount of space each data element takes up? Most measures will have more than 10 elements, which makes the "see at a glance" a bit more difficult. The more we can show without scrolling, the better.

DPC15038 commented:
![screen shot 2016-06-09 at 9 35 27 am](https://cloud.githubusercontent.com/assets/17532163/15931365/801d2318-2e25-11e6-8652-1a5c318ef023.png)

edeyoung commented:
Would we want this to be the default view or would we still want to show the very minimal view, and people can choose to display the additional information?

edeyoung commented:
and I agree, this looks really great!

edeyoung commented:
I'd also be curious to see what this looks like on an actual test case for measure testing. Maybe they don't use nearly as many fields so it's a much more manageable view.

thecristen commented:
Future user stories include letting the user be able to filter/hide/show certain types of elements, sort elements, etc. With that additional level control, I think having this info displayed by default is OK.

DPC15038 commented:
![example of sort by elements](https://cloud.githubusercontent.com/assets/17532163/16121108/f038361c-33af-11e6-8981-36194617e1b2.png)

DPC15038 commented:
![screen shot 2016-06-16 at 4 36 49 pm](https://cloud.githubusercontent.com/assets/17532163/16132266/6f6f988c-33e0-11e6-848a-5a4c9b906749.png)

DPC15038 commented:
![screen shot 2016-06-16 at 4 38 22 pm](https://cloud.githubusercontent.com/assets/17532163/16132322/a6ec4ed6-33e0-11e6-88f7-379a189c8f80.png)

DPC15038 commented:
![duration for data criteria](https://cloud.githubusercontent.com/assets/17532163/16385573/58906b9a-3c5b-11e6-8499-40ef75b41e0a.png)
Example of Duration

DPC15038 commented:
![code s now display when minimized](https://cloud.githubusercontent.com/assets/17532163/16385583/67c4bdb4-3c5b-11e6-8c68-75d471aac35b.png)
Codes display when minimized

DPC15038 commented:
![age now displays in patient builder](https://cloud.githubusercontent.com/assets/17532163/16385603/7b9bb450-3c5b-11e6-924a-8f4555f8d90f.png)
Age now displays in Patient Builder

edeyoung commented:
We recently got an email about needing to specify CPT codes in a very particular way in order to comply with a license agreement with AMA - so this may need to change. Don't know what that way is yet.

Also, it looks like the code numbers are no longer being displayed. I'd display the code number and the code description.

DPC15038 commented:
@edeyoung The code numbers are still being displayed when the data criteria is open, but not when it is closed. Should I change it to include, 1. Codeset, 2. Code Number, and 3. Code Description when open?

edeyoung commented:
Everything should display when open (code set, number, description). When closed, I think it should either display code set and number (to save space) or code set, number, and description. In all cases, the code number should be shown.

thecristen commented:
>  In all cases, the code number should be shown.

@edeyoung I'm not sure the code number should get such priority. We've heard from users that that number is more noise than signal. 

If we do include the number (when closed), maybe we can omit the code set - if the associated code set can be inferred from the length/format/value of the `number`.

edeyoung commented:
I know the Cypress folks were specifically interested in the description, but I didn't hear that they weren't interested in the code. Has this come up in other conversations also?

I'm concerned about omitting a piece of data that the users have become accustomed to seeing.

rutiah commented:
I agree with @edeyoung, the code can be helpful when debugging a test cases when the cause is value set related.

thecristen commented:
Please note that closed criteria showed *no information* on codes, so this isn't "omitting" anything that was previously available.

rutiah commented:
ah - wasn't thinking of the closed criteria. When open, it used to show codes only. I'm OK with not showing anything related to the code when the data criterion is closed. Usually I'm looking for attributes and date/time stamps.

edeyoung commented:
I think it's fine to not show anything about codes in the closed view if it doesn't make sense from a user perspective. If we are going to show code-related information, I think we should show the code number to maintain consistency and utility with what they're familiar with.

DPC15038 commented:
Here's a side-by-side of without code number and with code number. Thoughts?
<img width="1159" alt="side by side code comparison" src="https://cloud.githubusercontent.com/assets/17532163/16413356/71ffb3b6-3cff-11e6-87dd-682197d6e862.png">

edeyoung commented:
With the code. But I think Rute makes a good point - maybe we don't need the codes in the summary view at all.

What about losing the space between the codeset and the code?
e.g.: CPT:99365 - Initial comprehensive...

DPC15038 commented:
What do you think?? ![screen shot 2016-06-28 at 7 59 54 am](https://cloud.githubusercontent.com/assets/17532163/16414540/2927fafc-3d06-11e6-8544-13633a8cfed5.png)

edeyoung commented:
Looks good!!!

The "Show More" alignment seems to be a bit off.

DPC15038 commented:
Examples of Age spanning Measure Period<img width="565" alt="example of birth date" src="https://cloud.githubusercontent.com/assets/17532163/16461587/3c57fcc6-3dfb-11e6-8396-b04600f4e0fc.png">

DPC15038 commented:
Multiple **Show More** Buttons now format properly![multiple show mores](https://cloud.githubusercontent.com/assets/17532163/16461621/681b4610-3dfb-11e6-9202-d2b895a8876d.png)

DPC15038 commented:
Data Criteria Turns Red When Something is Not Performed 
![example of notperformed](https://cloud.githubusercontent.com/assets/17532163/16598882/f9474a7c-42cd-11e6-8831-ed9ea0148f33.png)

DPC15038 commented:
Changed the Way Age and Durations are Displayed
![screen shot 2016-07-05 at 4 31 45 pm](https://cloud.githubusercontent.com/assets/17532163/16598913/1669eac4-42ce-11e6-99e8-94fd4becdaf7.png)
----------
![screen shot 2016-07-05 at 4 31 41 pm](https://cloud.githubusercontent.com/assets/17532163/16598914/166a6b2a-42ce-11e6-907f-f5dfaf2522f8.png)

DPC15038 commented:
Changed the Red to Yellow http://puu.sh/pSnvJ/19d4c592b3.png

pkmitre commented on app/assets/javascripts/helpers.js.coffee:
```diff
@@ -42,14 +46,10 @@ Handlebars.registerHelper 'ifCond', (v1, operator, v2, options) ->
-  return out
+# Helper to retrieve the element at a given position in an array
+Handlebars.registerHelper 'lookup', (array, index) ->
+  return array[index] || undefined
```
If the array element at index is 0 or false, this will return undefined instead of the actual value. You could use the actual handlebars implementation (from the latest version of handlebars, which we don't currently use in Bonnie) from here: https://github.com/wycats/handlebars.js/blob/master/lib/handlebars/helpers/lookup.js

pkmitre replied:
Actually, looks like Lizzie already added that implementation to the patient dashboard branch, back in January of 2016...

pkmitre commented on app/assets/javascripts/helpers.js.coffee:
```diff
@@ -42,14 +46,10 @@ Handlebars.registerHelper 'ifCond', (v1, operator, v2, options) ->
+  return array[index] || undefined
+  
+# Helper to make use of moment date formatting
+Handlebars.registerHelper 'dateFormat', (date, fmt) ->
```
We already have a date formatting helper, called "moment", defined at the top of this file.

DPC15038 replied:
Back when I was making the modal for measure-upload-summary dialog for test-case-history there was a helper called lookup I was using, but that helper wasn't available in master, so i just went into test case history, and copied out the entire helper file and pasted it in my own

pkmitre commented on app/assets/javascripts/helpers.js.coffee:
```diff
@@ -19,8 +19,12 @@ Handlebars.registerHelper 'complexityIcon', (score) ->
   if bonnie.isPortfolio then options.fn(this) else options.inverse(this)
+  
+Handlebars.registerHelper "times", (n, block) ->
+  (block.fn(i) for i in [0...n]).join("")    
```
This implementation of times has different behavior from the one it's replacing in cases where n is 0 or otherwise falsy, the previous implementation appears to allow an else block to fire in those cases

DPC15038 commented:
Changed the location of the "Utility" Buttons, and added two new buttons
![screen shot 2016-07-18 at 2 33 27 pm](https://cloud.githubusercontent.com/assets/17532163/16925949/a56b3112-4cf4-11e6-8a95-75ec61f686fc.png)

pkmitre commented:
The new interface is definitely an improvement, but there are still a few things that feel odd about it from a design perspective, that I think make the interface feel less cohesive than it could be:

- The SHOW INFORMATION / HIDE INFORMATION button still feels like it's in an odd place; some controls are at the top of the PATIENT HISTORY and some are at the bottom of the ELEMENTS, what's the reasoning behind the placements? In particular, when there are a lot of elements, a user would have to scroll to find that button.

- There are three different controls that basically toggle between two states (SHOW INFORMATION / HIDE INFORMATION, OPEN ALL / CLOSE ALL, and ORDER BY DATA / ORDER BY ELEMENTS) and three different mechanisms that get used to do the toggling (a button that changes, two buttons to click that don't show state, and two buttons to click that do show state).

- Only the OPEN ALL and CLOSE ALL are behind the gear, it seems off that some items are behind a gear and some (ORDER BY) are not.

- What does ORDER BY "ELEMENTS" mean? Alphabetical order by name? If so, why not just say "NAME"?

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,181 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+  # Only allow the previewing/hiding to occur if they haven't clicked in .75 seconds
+  # TODO: The previewing/hiding button shares the same loading_spinner as the "Order By" buttons
+  # Maybe give it its own loading spinner?
+    if @timeSinceInfoPreviewClick + 750 < e.timeStamp
```
A better way to prevent multiple clicks is typically to disable the button as the first action after the click and then re-enable it once the action is complete (which you may get for free when the button changes). I think it's as easy as setting the disabled="true" property on the button, ie like @$('button#<idSelector>').prop('disabled', true).

pkmitre replied:
Also, why is the lag so long? There's a pretty noticeable lag even with just a few elements, which surprises me... I'll poke around a bit.

pkmitre commented on app/assets/javascripts/views/patient_builder/data_criteria.js.coffee:
```diff
@@ -79,6 +78,26 @@ class Thorax.Views.EditCriteriaView extends Thorax.Views.BuilderChildView
     @editFulfillmentHistoryView = new Thorax.Views.MedicationFulfillmentsView
       model: new Thorax.Model
       criteria: @model
+    @showPatientInformation = if @builderView.previousStateWasHideInfo() then 1 else 0
```
Two things when dealing with booleans:

1) avoid using `0` and `1`, use the `false` and `true` literals instead

2) avoid testing a condition and then assigning true or false as a result, ie for this particular line use the following:

    @showPatientInformation = @builderView.previousStateWasHideInfo()

pkmitre commented on app/assets/javascripts/views/patient_builder/data_criteria.js.coffee:
```diff
@@ -89,16 +108,54 @@ class Thorax.Views.EditCriteriaView extends Thorax.Views.BuilderChildView
+    fulfillments_information_array = []
+
+    if @showPatientInformation # only perform the calculations if the patient is previewing the information
+      if @model.attributes.fulfillments? # LOGIC FOR FULFILLMENTS
```
In Thorax, it's bad practice to refer to the attributes field directly, this should be 

    if @model.has('fulfillments')

And then, if you need to see the actual value someplace:

    @model.get('fulfillments')

pkmitre commented on app/assets/javascripts/views/patient_builder/data_criteria.js.coffee:
```diff
@@ -112,9 +169,18 @@ class Thorax.Views.EditCriteriaView extends Thorax.Views.BuilderChildView
+      references_information: references_information_array
+      result_information: results_information_array
+      fulfillment_information: fulfillments_information_array
+      duration_between_element_start_and_finish: element_duration
```
For many of these lines above: when the same variable appears in multiple contexts, it's generally a lot easier to follow the code if the variable name is the same in both contexts.

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,181 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+            sortByDateSelector.removeClass('active')
+            sortByDateSelector.html('DATE') # Remove the chevon arrow and bold tag from the other button, leave just the text
+            # Force these previous lines to occur before the sorting begins by waiting 0 seconds
+            setTimeout( =>
```
Can you explain a bit more about why the extensive use of setTimeout? It's a pretty unusual construct in view code (and maybe something we don't want to be doing), so I'd like to understand the reasoning more.

DPC15038 replied:
Hoss helped me with this bit, essentially I wanted the loading spinner to appear before anything else happened so the user could see that something was happening if/when the screen lagged. What was happening was that the loading spinner would appear after the lag started, so the screen would never update. Then when the lag stopped, the spinner would dissapear - AKA the user never sees the loading spinner and just sees lag. Hoss's solution was to set a timeout for 0 seconds, to force the code to show the spinner, then do everything else

pkmitre replied:
That makes sense, tweaking the comment could make this clearer, ie something like

    # Put JS code that takes a long time to run into a setTimeout so the UI thread is
    # free to update the display before processing starts

pkmitre commented on app/assets/javascripts/templates/patient_builder/edit_criteria.hbs:
```diff
@@ -21,6 +46,47 @@
+            {{#if fulfillment_information}}
+              <hr class= "hr-with-thin-margins"> {{! Thin line break between each element/value }}
+              {{#if isNegated}}<label>{{else}}<label class="control-label">{{/if}}<i class="fa fa-fw fa-medkit" aria-hidden="true"></i>Filled<input class="sr-only" disabled></label>
+                {{#each fulfillment_information}}
```
This #each looks like it's indented two spaces too far, which at first glance makes it look like it's controlled by the #if one line above it, which isn't the case.

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,181 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+        @$('#preview_information').text('Preview Information') # sets the button's new text
+        # Force these previous lines to occur before previewing patient info by waiting 0 seconds
+        setTimeout( =>
+          @trigger "hide_information_in_patient_builder"
```
The convention for event names is to use camel case with the first letter in lower case (hideInformationInPatientBuilder) rather than snake case.

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,181 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+    sortByElementsSelector = @$('#sort_by_elements')
+    sortByDateSelector = @$('#sort_by_date')
+    if e.target.id?
+      switch e.target.id
```
This is a pretty complex dance of looking at the ID of the element to figure out the action, a better way to do this would be to have two separate functions, one for each button (ie sortPatientEventsByDate and sortPatientEventsByName), that then call a common function for any common code.

thecristen replied:
I second the use two functions idea..

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,181 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+            sortByElementsSelector.addClass('active')
+            sortByElementsSelector.html('<i class="fa fa-chevron-right"></i> <b>ELEMENTS</b>') # add bold tag and arrow to more visibly show which button is selected
+            sortByDateSelector.removeClass('active')
+            sortByDateSelector.html('DATE') # Remove the chevon arrow and bold tag from the other button, leave just the text
```
Once you refactor this into two functions, it will probably also be cleaner to make use of jQuery selectors directly rather than assigning to variables, and perhaps chaining operators. ie

    @$('#sort_by_date').removeClass('active').html('DATE')

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,181 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+    # if the user accidentally clicks on DATE or ELEMENTS multiple times, it will sort multiple times
+    # even if everything is already sorted. By tracking sortButton, we can prevent the user from
+    # accidentally double clicking and waiting the extra few seconds that sorting takes.
+    loadingSpinnerSelector = @$('#loading_spinner')
```
The convention for HTML IDs is to use camel case with the first character lowercase, ie loadingSpinner

pkmitre commented:
The location of the age language seems a bit odd, it's wedged over to the right with the living status:

![screen shot 2016-07-19 at 10 31 11 am](https://cloud.githubusercontent.com/assets/3720400/16954259/39a9290c-4d9e-11e6-9058-c5e94ea532cc.png)

Seems like it would make more sense to include it under the birth date, since it's usually more closely related to that data, plus it makes better use of vertical real-estate:

![screen shot 2016-07-19 at 10 44 06 am](https://cloud.githubusercontent.com/assets/3720400/16954260/39aadee6-4d9e-11e6-9c10-ca8d0cf494c4.png)

pkmitre commented on app/assets/javascripts/templates/patient_builder/patient_builder.hbs:
```diff
@@ -108,9 +86,37 @@
+          <hr class="hr-with-thin-margins">
+          <h5><span class="patient-age-label">{{!The text that goes here is added via jQuery in patient_builder.js.coffee}}</span>
+            <span class="patient-age">{{patientAge}}</span>
+          <h5>
```
Closing h5 is missing /

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,181 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+      if @patientStatus.patientIsAlive
+        patientAgeLabelSelector.text("Age at Start of Measure Period:")
+      else
+        patientAgeLabelSelector.text("Age at Time of Death: ")
```
Setting text via jQuery isn't a good pattern to use within Thorax code, it should really be set by using {{}} in the view and passing the data via the context.

thecristen replied:
Note: that requires a full re-rendering in order to display a new value. So going with that approach, one should create a little Thorax view so that one could just re-render that view rather than the entire page.

pkmitre replied:
Good call, "microviews" can definitely be a win, depending on the size of the original view.

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,181 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+        patientAgeLabelSelector.text("Age at Start of Measure Period:")
+      else
+        patientAgeLabelSelector.text("Age at Time of Death: ")
+      patientAgeSelector.html("<br> #{@patientStatus.patientAge}") # Patient Age now displays all on one line (looks cleaner)
```
HTML (like <br>) should never appear in the view code, HTML should only appear in the template.

(Also see above re: setting text via jQuery)

pkmitre commented on app/assets/javascripts/models/patient.js.coffee:
```diff
@@ -130,8 +130,9 @@ class Thorax.Models.Patient extends Thorax.Model
     originalComparator = @get('source_data_criteria').comparator
-    @get('source_data_criteria').comparator = (crit) -> _(attributes).map((attr) -> crit.get(attr))
-    @get('source_data_criteria').sort()
+    for current_attribute in attributes.reverse() #.reverse() gives the highest priority to the beggining of the array
```
Variable names in JavaScript use the camel-case-with-first-letter-lowercase style, ie currentAttribute

pkmitre commented on app/assets/javascripts/models/patient.js.coffee:
```diff
@@ -130,8 +130,9 @@ class Thorax.Models.Patient extends Thorax.Model
-    @get('source_data_criteria').comparator = (crit) -> _(attributes).map((attr) -> crit.get(attr))
-    @get('source_data_criteria').sort()
+    for current_attribute in attributes.reverse() #.reverse() gives the highest priority to the beggining of the array
+      @get('source_data_criteria').comparator = current_attribute #.comparator cannot act on an array, it must be a single element
```
I don't believe the statement "comparator cannot act on an array, it must be a single element" is true, I tested out the approach used in the original code and it works fine; it's also more efficient than this approach of sorting multiple times.

pkmitre replied:
To be clear: to sort on an array, the array has to be provided in the context of a sort function, not as a literal array, ie

    @get('source_data_criteria').comparator = [attribute1, attribute2]

won't work but

    @get('source_data_criteria').comparator = (model) -> [model.get('attribute1'), model.get('attribute2')]

will.

DPC15038 replied:
> comparator cannot act on an array, it must be a single element
This might be false, but it is true that the original code doesn't work under all circumstances. It can sort by name, however, it cannot sort dates under 2000

pkmitre replied:
Why can't it sort dates under 2000? The following two pieces of code seem like they should be equivalent:

    @get('source_data_criteria').comparator = 'start_date'

    @get('source_data_criteria').comparator = (model) -> [model.get('start_date')]

You're saying that the second version handles dates differently?

DPC15038 replied:
Working on a solution now...
![image](https://cloud.githubusercontent.com/assets/17532163/17106518/aff1c51e-5259-11e6-95d6-36ef55b9be34.png)

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,181 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+            sortByElementsSelector.html('ELEMENTS') # Remove the chevon arrow and bold tag from the other button, leave just the text
+            # Force these previous lines to occur before the sorting begins by waiting 0 seconds
+            setTimeout( =>
+              @model.sortCriteriaBy 'start_date'
```
Elsewhere where we sort by date we also include the end_date as a secondary sort option.

DPC15038 replied:
Yep! If my understanding is correct, secondary sort options only occur when the primary ones are equal. I'm not entirely sure how comparator and .sort() work, but even without the secondary option of 'end_date' two elements with the same start date automatically sort by this secondary condition. As a result, I intentionally didn't add that secondary sort option

pkmitre replied:
I suspect that what you saw was just a coincidence, since if you don't make it explicit what would actually cause that to happen? I just tested it, and it doesn't actually sort by the end_date when the start_dates are equal.

pkmitre commented on app/assets/javascripts/templates/patient_builder/edit_criteria.hbs:
```diff
@@ -1,5 +1,10 @@
+  <div class="criteria-type-marker">
+  <i class="fa fa-fw fa-lg {{faIcon}}" aria-hidden="true"></i>
+{{/if}}
 </div>
```
The only thing that's different in the two code blocks is the class, so this could be simplified as

    <div class="{{#if isNegated}}criteria-type-marker-with-negation{{else}}criteria-type-marker{{/if}}
      <i class="fa fa-fw fa-lg {{faIcon}}" aria-hidden="true"></i>
    </div>

pkmitre commented on app/assets/javascripts/views/patient_builder/data_criteria.js.coffee:
```diff
@@ -89,16 +108,54 @@ class Thorax.Views.EditCriteriaView extends Thorax.Views.BuilderChildView
+             return ((ful.get('quantity_dispensed_value') + " " + ful.get('quantity_dispensed_unit') + " " + ful.get('dispense_date') + " ").replace(/_/g, ' ').replace(/\w\S*/g, (txt) ->  txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase()) + ful.get('dispense_time')))
+
+      if @model.attributes.field_values? # LOGIC FOR FIELD VALUES
+        that_model = @model #store this.model to be used inside the map
```
The reason you can't access @model inside the map is because you're using a thin arrow for the function, use a fat arrow (=>) and you won't need to juggle with a variable called "that_model". To read more about it look for the section called "Bound Functions" in http://coffeescript.org/

pkmitre commented on app/assets/javascripts/views/patient_builder/data_criteria.js.coffee:
```diff
@@ -112,9 +169,18 @@ class Thorax.Views.EditCriteriaView extends Thorax.Views.BuilderChildView
       faIcon: @model.faIcon()
       definition_title: definition_title
       canHaveNegation: @model.canHaveNegation()
+      isNegated: @model?.get('negation')
```
Why the existential operator (?) here? I don't think it's needed.

pkmitre commented on app/assets/javascripts/views/patient_builder/data_criteria.js.coffee:
```diff
@@ -204,6 +272,12 @@ class Thorax.Views.EditCriteriaView extends Thorax.Views.BuilderChildView
 
+  getNegationCodeIDAsText: (value_set_oid) ->
+    if @model.measure()?.valueSets().where(oid: value_set_oid).length > 0
+      return @model.measure()?.valueSets().where(oid: value_set_oid)[0].attributes.display_name
```
I think we can drop the existential operator here too, particularly since if it's used the expressions will still blow up because the chaining continues a few steps beyond (ie if there's no measure() the where() will blow up)

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -15,7 +15,7 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
     @editCriteriaCollectionView = new Thorax.CollectionView
       collection: @model.get('source_data_criteria')
-      itemView: (item) => new Thorax.Views.EditCriteriaView(model: item.model, measure: @measure)
+      itemView: (item) => new Thorax.Views.EditCriteriaView(model: item.model, measure: @measure, builderView: @)
```
Because EditCriteriaView is a child of BuilderView, you shouldn't need to pass builderView in, you should be able to access it via the @patientBuilder() function.

thecristen commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,181 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+
+    if @patientStatus.patientAge == null # Meaning the Birthdate came after the deathdate or the measurePeriod
+      if @patientStatus.patientIsAlive # Changes Warning Message based on whether the patient is alive or not
+        patientAgeLabelSelector.removeClass('fail fa fa-fw fa-times-circle')
```
removing and adding the `fa` and `fa-fw` classes is redundant; you can leave them out of this code and the classes will be left alone.

thecristen commented on app/assets/stylesheets/builder.less:
```diff
@@ -422,3 +439,24 @@
+}
+
+.btn-width-100-percent {
+  width:100%;
```
There's a class in Bootstrap that does this! :D It's `.btn-block`.

thecristen commented on app/assets/stylesheets/mixins.less:
```diff
@@ -20,6 +20,9 @@
   color: @brand-primary;
 }
 
+.warning {
```
You could probably eliminate this and use Bootstrap's `.text-warning` class to achieve this.

DPC15038 replied:
The top is with .text-warning, and the bottom is what I currently have. Thoughts?
![screen shot 2016-07-22 at 8 39 44 am](https://cloud.githubusercontent.com/assets/17532163/17057453/ef735960-4fe7-11e6-8e85-3d86bf5df9eb.png)

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -77,7 +82,7 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
     rendered: ->
       @$('.draggable').draggable revert: 'invalid', helper: 'clone', appendTo: '.patient-builder', zIndex: 10
-
+      @setPatientAge() #Clicking the button that removes Death Date doesn't trigger materialize, it triggers "render"
```
It seems like the lack of a materialize when removing the death date is an issue, shouldn't it materialize?

thecristen commented on app/assets/javascripts/templates/patient_builder/patient_builder.hbs:
```diff
@@ -142,16 +148,47 @@
+        <span class="sr-only">Patient Builder Utilities</span>
+      {{/button}}
+      {{#button "openAllDataCriteria" id="open_all" class="btn btn-sm btn-primary hidden"}}Open All
+        <span class="sr-only">Open all Data Criteria</span>
```
FYI, the screen reader will read both the visible "Open all" and the hidden "Open all" from "Open all Data Criteria". We probably just need the sr-only span to say "data criteria"

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -139,6 +144,7 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
 
   materialize: ->
     @serializeWithChildren()
+    @setPatientAge() # When a birthdate or deathdate is selected, materialize() is called and sets the Patient's Age
```
Rather than hooking into materialize(), which happens every time any change is made to the patient record, it might make more sense to just set the patient age specifically when the birth date changes.

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,181 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+      elementDurationAsString = elementDurationAsString.slice(0,-2) # Chop off the trailing ", " from the string
+    else
+      elementDurationAsString = null # meaning the end date preceded the start date
+    return elementDurationAsString
```
This code could be much simpler. If this is the behavior we want, we could probably start with something like

    getDuration: (start_date, end_date) ->
      duration = moment.duration(end_date.diff(start_date))
      return null if duration < 0
      terms = []
      for unit in ['years', 'months', 'days', 'hours', 'minutes']
        terms.push "#{duration.get(unit)} #{unit}" if duration.get(unit) > 0
        break if terms.length > 1
      return terms.join(', ')

(not fully tested)

If we're willing to simplify the output, we could even just use the humanize() method that moment.js provides for durations.

pkmitre replied:
Also, it would be nice if we had the pluralization correctly match the number, ie 1 year vs 2 years

DPC15038 replied:
It currently has this feature

pkmitre replied:
Missed that! My suggested code would still need to be tweaked...

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,181 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+        else
+          # Time picker wants to offset birthdate by 5 hours - not sure if problem with timepicker or Moment.js
+          # either way, adding 5 hours to birthdate fixes problem
+          @patientStatus.patientAge = @getDuration(moment(@model.get('birthdate') * 1000).add(5, 'hours'), moment(new Date(bonnie.measurePeriod, 0,1,0,0,0,0)))
```
This is likely due to a timezone difference, ie EST vs GMT; the right way to handle that is to use moment.utc (search for that in the patient builder code).

DPC15038 replied:
I did some testing and here are the results

![screen shot 2016-07-21 at 1 10 31 pm](https://cloud.githubusercontent.com/assets/17532163/17031908/837ac908-4f44-11e6-86b1-7916377b79a5.png)

pkmitre replied:
There's still a fundamental issue with the approach of "adding 5 seems to fix the problem". If Bonnie is installed in a different timezone, will this still work? We need to figure out what's going on and figure out a solution that doesn't use a "magic number".

DPC15038 replied:
Ooh I forgot about timezones, I'll look more into this, and come up with a dynamic solution

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,181 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+      elementDurationAsString = null # meaning the end date preceded the start date
+    return elementDurationAsString
+
+  getPatientAge: ->
```
The approach used in the method of building a big global status object to hold patient status that gets used elsewhere has a few issues (it's a lot like a global variable, it doesn't separate concerns, there's computation being done on model data in the view that would be better done in the model). A better approach would be to create individual functions on the patient model like isAlive() and age(), and just call those functions where needed (ie passing the context from the view to the template). Most of the code should be in the model and the layout, with likely only a bit of code in the view.

DPC15038 commented:
Having weird computer issues, so pushing up changes now - No need to review these currently, many things still incomplete

pkmitre commented on app/assets/javascripts/models/data_criteria.js.coffee:
```diff
@@ -35,7 +35,7 @@ class Thorax.Collections.MeasureDataCriteria extends Thorax.Collection
 class Thorax.Models.PatientDataCriteria extends Thorax.Model
   idAttribute: null
-  
+
```
It looks like the only changes in this file are white space.

DPC15038 replied:
One of my Atom packages auto removes trailing whitespace on save. Is this a problem? Should I revert the file?

pkmitre replied:
It makes code reviews more difficult; revert this file (since there are no actual code changes) and adjust atom to not automatically remove trailing white space

pkmitre commented on app/assets/javascripts/models/patient.js.coffee:
```diff
@@ -29,7 +29,18 @@ class Thorax.Models.Patient extends Thorax.Model
     new @constructor JSON.parse(json), parse: true
 
-  getBirthDate: -> new Date(@get('birthdate'))
+  getBirthDate: -> moment(moment(moment.unix(@get('birthdate'))).utc().format("YYYY-MM-DD HH:mm"))
```
This seems somewhat convoluted, what's the goal of this line?

DPC15038 replied:
Formats the Date for Moment JS, and accounts for the weird +5 hour timezone issue I found earlier. I'll see if I can refactor it a bit

pkmitre commented on app/assets/javascripts/models/patient.js.coffee:
```diff
@@ -29,7 +29,18 @@ class Thorax.Models.Patient extends Thorax.Model
+    if @get('expired')
+      false
+    else
+      true
```
This could be simplified to

    !@get('expired')

pkmitre commented on app/assets/javascripts/models/result.js.coffee:
```diff
@@ -188,7 +188,7 @@ class Thorax.Models.Result extends Thorax.Model
   calculateDataCriteriaOrCounts: (rationale) ->
     orCounts = {}
-    for key, dc of @measure.get('data_criteria') when dc.derivation_operator == 'UNION' && key.indexOf('UNION') != -1
+    for key, dc of @measure.get('data_criteria') when dc.derivation_operator == 'UNION' && key.match(/UNION|satisfiesAny/)
```
I'm pretty sure this change is on master, we should look into why it's showing up in the diff for your pull request

DPC15038 replied:
I rebased with master

pkmitre replied:
If code you didn't write is showing up in your pull request that means there's an issue with how you rebased; we can take a look later today

pkmitre commented on app/assets/javascripts/utilities.js.coffee:
```diff
@@ -0,0 +1,13 @@
+window.Bonnie ||= {}
+Bonnie.util ||= {}
+Bonnie.util.getDurationBetween = (start_date, end_date) ->
```
We should check the case and be consistent, I think I've seen Bonnie and bonnie

pkmitre commented on app/assets/javascripts/views/patient_builder/data_criteria.js.coffee:
```diff
@@ -79,6 +78,24 @@ class Thorax.Views.EditCriteriaView extends Thorax.Views.BuilderChildView
+      @render() # (Re)rendering the view causes each specific data criteria to close
+    )
+    @listenTo(@parentView, "openAllDataCriteria", ->
+      e = arguments[0] # e is only passed within the arguments[] object
```
This should just be

    @listenTo(@parentView, "openAllDataCriteria", (e) ->

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -238,6 +244,58 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
+    if @$('#previewPatientInformation').is(":checked")
+      true
+    else
+      false
```
We should avoid using the DOM to store application state.

(Also, avoid explicit references to true and false)

pkmitre commented on app/assets/javascripts/models/data_criteria.js.coffee:
```diff
@@ -96,6 +96,12 @@ class Thorax.Models.PatientDataCriteria extends Thorax.Model
 
+  getNegationCodeText: ->
+    if @get('negation')
+      @.measure().valueSets().where(oid: @get('negation_code_list_id'))[0].get('display_name')
```
Should be @measure instead of @.measure

pkmitre commented on app/assets/javascripts/models/patient.js.coffee:
```diff
@@ -29,7 +29,15 @@ class Thorax.Models.Patient extends Thorax.Model
 
-  getBirthDate: -> new Date(@get('birthdate'))
+  # Returns the BirthDate formatted as a moment from Moment JS
+  getBirthDate: -> moment(moment.unix(@get('birthdate')).utc().format("YYYY-MM-DD HH:mm"))
```
I think this can be simplified, take a look at the way moment is used elsewhere in the code to deal with UTC and unix time. I think this could just be something like

    moment.utc(@get('birthdate'), 'X')

If it's not quite right grab me and we can sit down and figure out what does make sense.

pkmitre commented on app/assets/javascripts/models/patient.js.coffee:
```diff
@@ -29,7 +29,15 @@ class Thorax.Models.Patient extends Thorax.Model
+    if @isAlive()
+      null
+    else
+      return moment(moment(moment.unix(@get('deathdate'))).utc().format("YYYY-MM-DD HH:mm"))
```
See moment comment above.

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -15,7 +15,7 @@ class Thorax.Views.PatientBuilder extends Thorax.Views.BonnieView
     @editCriteriaCollectionView = new Thorax.CollectionView
       collection: @model.get('source_data_criteria')
-      itemView: (item) => new Thorax.Views.EditCriteriaView(model: item.model, measure: @measure)
+      itemView: (item) => new Thorax.Views.EditCriteriaView(model: item.model, measure: @measure, parentView: @)
```
We shouldn't need to pass the parent view. `EditCriteriaView` is an instance of `BuilderChildView`, which has a method called `patientBuilder()`

pkmitre commented on app/assets/javascripts/views/patient_builder/data_criteria.js.coffee:
```diff
@@ -79,6 +79,23 @@ class Thorax.Views.EditCriteriaView extends Thorax.Views.BuilderChildView
     @editFulfillmentHistoryView = new Thorax.Views.MedicationFulfillmentsView
       model: new Thorax.Model
       criteria: @model
+    @previewElementInformation = @parentView.previewElementInformation # Allows us to remember state after sorting
```
This should just be set in the context rather than as an instance variable... that simplifies the code below quite a bit, to

    @listenTo @parentView, "previewInformationinDataCriteria", => @render()

pkmitre commented on app/assets/javascripts/views/patient_builder/data_criteria.js.coffee:
```diff
@@ -79,6 +79,23 @@ class Thorax.Views.EditCriteriaView extends Thorax.Views.BuilderChildView
+    @listenTo(@parentView, "previewInformationinDataCriteria", ->
+      @previewElementInformation = true
+      @render()
+    )
```
The coffeescript style we use avoids parentheses in these cases, ie

    @listenTo @parentView, "previewInformationinDataCriteria", =>
        @previewElementInformation = true
        @render()

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -251,3 +304,49 @@ class Thorax.Views.BuilderPopulationLogic extends Thorax.LayoutView
+  template: JST['patient_builder/patient_age']
+
+  initialize: ->
+    #We shouldn't need to listen for the change event, but Thorax doesn't seem to be updating the model properly
```
Our code style includes a space between the # and the text, ie

    # We shouldn't need to listen for the change event, but Thorax doesn't seem to be updating the model properly

pkmitre replied:
Also, the comment isn't quite right, it should read more like

    # We shouldn't need to listen for the change event, but Thorax doesn't seem to be updating the view given the change event

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -251,3 +304,49 @@ class Thorax.Views.BuilderPopulationLogic extends Thorax.LayoutView
+
+  initialize: ->
+    #We shouldn't need to listen for the change event, but Thorax doesn't seem to be updating the model properly
+    @model.on 'change', =>  @render()
```
It's worth being consistent with spacing, ie

    @model.on 'change', => @render()

instead of

    @model.on 'change', =>  @render()

It doesn't have a big impact here, but there are definitely cases where it affects readability

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -251,3 +304,49 @@ class Thorax.Views.BuilderPopulationLogic extends Thorax.LayoutView
+  context: ->
+    if @model.isAlive()
+    # if the birthdate/deathdate fields are left blank they default to the epoch
+      unless @model.getBirthDate().isSame(moment("1970-01-01 00:00"))
```
I'm assuming that `getBirthdate()` is returning this date if it's `null`? That's not a great practice, what about a patient that actually has that birthdate? If the birthdate is `null`, `getBirthdate()` should probably return `null`.

pkmitre commented on app/assets/javascripts/views/patient_builder/patient_builder.js.coffee:
```diff
@@ -251,3 +304,49 @@ class Thorax.Views.BuilderPopulationLogic extends Thorax.LayoutView
+          patientAge = "Age at Start of Measure Period: #{Bonnie.util.getDurationBetween(@model.getBirthDate(), moment([bonnie.measurePeriod]))}"
+    else
+      unless @model.getBirthDate().isSame(moment("1970-01-01 00:00")) || @model.getDeathDate().isSame(moment("1970-01-01 00:00"))
+        if Bonnie.util.getDurationBetween(@model.getBirthDate(), @model.getDeathDate())
```
This condition uses an implicit property of the `getDurationBetween()` method (ie returns null if the duration would be negative), which makes the code trickier to understand than it needs to be. Why not

    if @model.getBirthDate() < @model.getDeathDate()

DPC15038 commented:
@pkmitre I'd like to sit down with you and go over some of your comments/ask a few questions before you take another look

thecristen commented:
"End Date precedes Start Date" seems to be shown on a data element even when the end date is set equal to the start date

edeyoung commented:
I think the measure request to include icons for outdated codes might be a good opportunity to merge this PR in.

I think some of the UI should change slightly. See my screen shot below and the specific changes I'm interested in:

![image](https://cloud.githubusercontent.com/assets/11574199/25233850/383f9806-25ae-11e7-9e9c-ebc547f31b58.png)

1. Make "Order By" a drop down box in case we want to add more options. Also, less visual clutter.
2. Remove "collapse all" and "expand all". With the preview information, this seems like less of a necessity.
3. Change wording and placement of the "show details" check box.
4. Include codes in the detail view.
5. (in expanded view) include time duration

DPC15038 commented:
Oh god did I just include all the commits from everything...Should I cherry pick just mine, and worry about merging/git issues later?

c-monkey commented:
I think cherry picking does seem wise!

DPC15038 commented:
Cherrypicked the correct commits, and force pushed to remove any unwanted commits in this PR.
